### PR TITLE
FXC-5121: Add option to drop filtered modes

### PR DESCRIPTION
**Pull request by momchil-flex**
- Introduces `drop_modes` boolean, if `True`, just drop modes that don't make the filtering threshold. The caveat here is that the number of modes might depend on the frequency. The choice that was made was to do the dropping *after* the tracking, and then just do the intersection and keep the smallest number of modes that meet the filtering criterion at every frequency.
- Introduces `fill_fraction_box` as a filtering key which then requires a `bounding_box` to be passed to the `sort_spec`. Also a `fill_fraction(box)` method to the mode solver data, as well as a property `fill_fraction_box` that uses the box defined in the stored `mode_spec.sort_spec`.

---

> [!NOTE]
> Introduces configurable mode retention and a spatial fill-fraction metric for mode filtering/sorting.
> 
> - Adds `ModeSortSpec.keep_modes` ("all" | "filtered" | integer N) to keep all, only filtered, or top N modes; validates against `num_modes` and requires `filter_key` when "filtered"
> - Adds `fill_fraction_box` metric with `ModeSortSpec.bounding_box`; implements `ModeSolverData.fill_fraction(box)` and `fill_fraction_box` accessor with planar/intersection validation
> - Extends `ModeSolverData.sort_modes()` to apply filter/sort, track across frequencies, then drop/limit modes and update `num_modes`; utility `_apply_mode_subset()` added
> - Updates JSON schemas and defaults to include `bounding_box` and `keep_modes`; expands MODE_DATA_KEYS
> - Adds comprehensive tests for new behavior, edge cases, and validations; updates CHANGELOG
> - CI: allow schema changes on pre-release (rc/dev) base versions in workflow
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c3d551070b5f44f71ad2d7721f7ce8c5a19d80e5. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h2>Greptile Overview</h2>

<h2>Greptile Overview</h2>

### Greptile Summary

This PR introduces two major enhancements to the mode solver functionality:

## Key Changes

**1. Mode Retention Control (`keep_modes`)**
- Adds `ModeSortSpec.keep_modes` field accepting "all" (default), "filtered", or integer N
- When set to "filtered", only modes passing the filter criterion at ALL frequencies are retained
- When set to integer N, keeps top N modes after sorting/filtering
- Implements `_apply_mode_subset()` method to efficiently subset mode data arrays
- Includes proper validation: "filtered" requires `filter_key`, integer N must be ≤ `num_modes`

**2. Spatial Fill-Fraction Metric (`fill_fraction_box`)**
- Adds `fill_fraction_box` as a new MODE_DATA_KEYS literal for sorting/filtering
- Requires `ModeSortSpec.bounding_box` parameter (validated at construction)
- Implements `fill_fraction(box)` method computing field-energy ratio inside/outside box
- Implements `fill_fraction_box` property as convenience accessor using `sort_spec.bounding_box`
- Includes proper planar monitor validation and intersection checks

## Implementation Quality

**Strengths:**
- Comprehensive test coverage including edge cases (all modes filtered, validation errors)
- Proper floating-point comparisons using `np.isclose`/`np.allclose`
- Clear error messages using single quotes for variable names (per style guide)
- Proper separation of filtering (before tracking) and mode retention (after tracking)
- Efficient array manipulation using `np.take_along_axis` for mode subsetting

**Issues Found:**
- Minor: `fill_fraction_box` property should use `@cached_property` instead of `@property` since it performs expensive computation

## Architecture

The implementation follows a clear pipeline:
1. Filter and sort modes at each frequency independently
2. Apply mode tracking across frequencies if requested  
3. Re-evaluate filter criteria after tracking (for consistency)
4. Subset modes based on `keep_modes` parameter
5. Update `num_modes` metadata to reflect final mode count

The design correctly handles the caveat mentioned in the PR description: mode count may vary by frequency, resolved by taking the intersection of filtered modes across all frequencies.

### Confidence Score: 4/5

- This PR is generally safe to merge with minor style improvements recommended
- Score reflects well-tested implementation with comprehensive validation, but one style issue (property decorator) should be addressed. The complex sorting/filtering/tracking logic appears sound after thorough analysis. Edge cases are properly handled with appropriate error messages. No critical bugs or security issues found.
- Pay attention to tidy3d/components/data/monitor_data.py for the property decorator style issue at line 807

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| CHANGELOG.md | 5/5 | Changelog entries are clear and follow conventions. Properly documents new features 'keep_modes' and 'fill_fraction_box' with specific names as required. |
| tidy3d/components/mode_spec.py | 4/5 | Adds 'keep_modes' field and 'bounding_box' to ModeSortSpec with proper validation. Validators ensure 'filtered' requires 'filter_key' and 'fill_fraction_box' requires 'bounding_box'. Minor style issue with property decorator. |
| tidy3d/components/data/monitor_data.py | 3/5 | Implements fill_fraction calculations and mode subset logic. Complex sorting/filtering/tracking logic appears sound. One style issue: 'fill_fraction_box' property should use '@cached_property' for performance. Warning message formatting follows guidelines. |
| tests/test_plugins/test_mode_solver.py | 4/5 | Comprehensive tests for new features including 'keep_modes', 'fill_fraction_box', edge cases (all modes filtered, validation errors). Tests use proper assertions, parametrization, and floating-point comparisons with 'np.isclose'/'np.allclose'. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant ModeSolverData
    participant ModeSortSpec
    participant sort_modes
    participant _apply_mode_subset
    
    User->>ModeSortSpec: Create with keep_modes="filtered"
    ModeSortSpec->>ModeSortSpec: Validate filter_key is set
    ModeSortSpec->>ModeSortSpec: Validate bounding_box if fill_fraction_box used
    
    User->>ModeSolverData: Call sort_modes(sort_spec)
    ModeSolverData->>sort_modes: Enter sort_modes method
    
    alt filter_key is fill_fraction_box
        sort_modes->>ModeSolverData: Call fill_fraction_box property
        ModeSolverData->>ModeSolverData: Compute fill_fraction(bounding_box)
        ModeSolverData-->>sort_modes: Return fill fraction metric
    end
    
    sort_modes->>sort_modes: For each frequency: filter and sort modes
    sort_modes->>sort_modes: Apply mode tracking if requested
    
    alt keep_modes == "filtered"
        sort_modes->>sort_modes: Re-evaluate filter after tracking
        sort_modes->>sort_modes: Compute keep_mask across all frequencies
        alt No modes pass filter
            sort_modes-->>User: Raise ValidationError
        else Some modes pass
            sort_modes->>sort_modes: Set keep_inds = where(keep_mask)
        end
    else keep_modes is integer N
        sort_modes->>sort_modes: Set keep_inds = arange(N)
        alt N > num_modes
            sort_modes-->>User: Raise ValidationError
        end
    end
    
    alt keep_inds is not None
        sort_modes->>_apply_mode_subset: Subset to keep_inds
        _apply_mode_subset->>_apply_mode_subset: Use take_along_axis to extract modes
        _apply_mode_subset-->>sort_modes: Return subset data
        sort_modes->>ModeSolverData: Update monitor.mode_spec.num_modes
    end
    
    sort_modes-->>User: Return filtered/sorted ModeSolverData
```

<h3>Important Files Changed</h3>

File Analysis

| Filename | Score | Overview |
|----------|-------|----------|
| CHANGELOG.md | 5/5 | Changelog entries are clear and follow conventions. Properly documents new features 'keep_modes' and 'fill_fraction_box' with specific names as required. |
| tidy3d/components/mode_spec.py | 4/5 | Adds 'keep_modes' field and 'bounding_box' to ModeSortSpec with proper validation. Validators ensure 'filtered' requires 'filter_key' and 'fill_fraction_box' requires 'bounding_box'. Minor style issue with property decorator. |
| tidy3d/components/data/monitor_data.py | 3/5 | Implements fill_fraction calculations and mode subset logic. Complex sorting/filtering/tracking logic appears sound. One style issue: 'fill_fraction_box' property should use '@cached_property' for performance. Warning message formatting follows guidelines. |
| tests/test_plugins/test_mode_solver.py | 4/5 | Comprehensive tests for new features including 'keep_modes', 'fill_fraction_box', edge cases (all modes filtered, validation errors). Tests use proper assertions, parametrization, and floating-point comparisons with 'np.isclose'/'np.allclose'. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant ModeSolverData
    participant ModeSortSpec
    participant sort_modes
    participant _apply_mode_subset
    
    User->>ModeSortSpec: Create with keep_modes="filtered"
    ModeSortSpec->>ModeSortSpec: Validate filter_key is set
    ModeSortSpec->>ModeSortSpec: Validate bounding_box if fill_fraction_box used
    
    User->>ModeSolverData: Call sort_modes(sort_spec)
    ModeSolverData->>sort_modes: Enter sort_modes method
    
    alt filter_key is fill_fraction_box
        sort_modes->>ModeSolverData: Call fill_fraction_box property
        ModeSolverData->>ModeSolverData: Compute fill_fraction(bounding_box)
        ModeSolverData-->>sort_modes: Return fill fraction metric
    end
    
    sort_modes->>sort_modes: For each frequency: filter and sort modes
    sort_modes->>sort_modes: Apply mode tracking if requested
    
    alt keep_modes == "filtered"
        sort_modes->>sort_modes: Re-evaluate filter after tracking
        sort_modes->>sort_modes: Compute keep_mask across all frequencies
        alt No modes pass filter
            sort_modes-->>User: Raise ValidationError
        else Some modes pass
            sort_modes->>sort_modes: Set keep_inds = where(keep_mask)
        end
    else keep_modes is integer N
        sort_modes->>sort_modes: Set keep_inds = arange(N)
        alt N > num_modes
            sort_modes-->>User: Raise ValidationError
        end
    end
    
    alt keep_inds is not None
        sort_modes->>_apply_mode_subset: Subset to keep_inds
        _apply_mode_subset->>_apply_mode_subset: Use take_along_axis to extract modes
        _apply_mode_subset-->>sort_modes: Return subset data
        sort_modes->>ModeSolverData: Update monitor.mode_spec.num_modes
    end
    
    sort_modes-->>User: Return filtered/sorted ModeSolverData
```

### CODE REVIEW

**Comment by caseyflex** [comment]
@momchil-flex thanks for this!

I think in terms of API, I would prefer a field `keep_num_modes` or something, so that after the filtering and sorting is applied we can just keep the top `keep_num_modes`. Instead of `drop_modes`. The default can be `None` to keep all modes.

If you want I can take over this PR and make the change

**Comment by momchil-flex** [comment]
I like the suggestion in that the mode spec would still always produce a deterministic number of modes.

But in terms of functionality, doesn't it sound not adequate e.g. for the EME filtering? Like if you strictly want to drop modes that don't meet a given criterion, this is not going to achieve that, or would achieve in some cases by chance?

Maybe you're imagining having a separate filtering done in EME, which I think could make sense.

**Comment by caseyflex** [comment]
> I like the suggestion in that the mode spec would still always produce a deterministic number of modes.
> 
> But in terms of functionality, doesn't it sound not adequate e.g. for the EME filtering? Like if you strictly want to drop modes that don't meet a given criterion, this is not going to achieve that, or would achieve in some cases by chance?
> 
> Maybe you're imagining having a separate filtering done in EME, which I think could make sense.

It’s the closest you can easily get to deterministic number of modes, although not perfect. The sim coeffs can tell you how many modes were actually kept; if it’s fewer than the number requested, then you may need to increase num_modes for the computation, or loosen the filter

**Comment by momchil-flex** [comment]
Ok yea go for it.

**Comment by caseyflex** [comment]
> Ok yea go for it.

Maybe a field “keep_modes” which can be “all”, “filtered”, or an integer

**Review by caseyflex — APPROVED** [review]
@momchil-flex I kept support for your method via `keep_modes="filtered"`.

I also added the option to have `keep_modes` be an integer, in which case it just keeps that many modes regardless of filtering, warning however if it keeps some modes that didn't pass the filer.

**Comment by momchil-flex** [comment]
Sorry, accidentally closed this. Sounds good for your change, I guess we'll work on finalizing this after the holidays but good for testing for now!

File: tidy3d/components/data/monitor_data.py [hunk 1]
```diff
@@ -2404,31 +2572,83 @@ def _order_indices(indices, vals_all):
...
+            for ifreq in range(num_freqs):
+                vals = filter_metric_sorted.isel(f=ifreq).values
+                if sort_spec.filter_order == "over":
+                    mask = vals >= sort_spec.filter_reference
+                else:
+                    mask = vals <= sort_spec.filter_reference
+                masks_after.append(mask)
+
+            keep_mask = np.all(np.stack(masks_after, axis=0), axis=0)
+        if keep_modes == "filtered":
```
[the remark this answers is not in the document]
Comment (caseyflex) [reply tidy3d/components/data/monitor_data.py]: not correct but I'll restructure to be more clear

File: tidy3d/components/data/monitor_data.py [hunk 2]
```diff
@@ -2404,31 +2572,83 @@ def _order_indices(indices, vals_all):
...
+            if not np.any(keep_mask):
+                raise ValidationError(
+                    "Filtering removes all modes; relax the filter threshold or change 'keep_modes'."
+                )
+
+            num_modes_sorted = filter_metric_sorted.sizes["mode_index"]
+            if keep_mask.sum() < num_modes_sorted:
+                keep_inds = np.where(keep_mask)[0]
+        elif isinstance(keep_modes, int):
```
[the remark this answers is not in the document]
Comment (caseyflex) [reply tidy3d/components/data/monitor_data.py]: same as above

File: tidy3d/components/data/monitor_data.py [hunk 3]
```diff
@@ -2404,31 +2572,83 @@ def _order_indices(indices, vals_all):
...
+            if keep_mask.sum() < keep_modes:
+                log.warning(
+                    f"Filtering with 'keep_modes={keep_modes}' keeps "
+                    f"more than the '{keep_mask.sum()}' modes which pass the filter. Consider "
+                    "relaxing the filter threshold or changing 'keep_modes'."
+                )
+            keep_inds = np.arange(keep_modes)
```
[the remark this answers is not in the document]
Comment (caseyflex) [reply tidy3d/components/data/monitor_data.py]: same issue. The logic was correct, but indenting it makes it clearer so I'll do it

**Comment by caseyflex** [comment]
@greptile

File: tidy3d/components/data/monitor_data.py [hunk 4]
```diff
@@ -2404,31 +2570,83 @@ def _order_indices(indices, vals_all):
...
         if track_freq and num_freqs > 1:
             data_sorted = data_sorted.overlap_sort(track_freq)
 
+        keep_inds = None
+        if keep_modes == "filtered" or isinstance(keep_modes, int):
+            # Re-evaluate the filter after sorting/tracking so modes are dropped consistently.
+            if sort_spec.filter_key == "fill_fraction_box":
+                filter_metric_sorted = data_sorted.fill_fraction_box
+            else:
+                filter_metric_sorted = getattr(data_sorted, sort_spec.filter_key)
```
[the remark this answers is not in the document]
Comment (caseyflex) [reply tidy3d/components/data/monitor_data.py]: this won't run if keep_modes is "all"

File: tidy3d/components/data/monitor_data.py [hunk 5]
```diff
@@ -2404,31 +2570,83 @@ def _order_indices(indices, vals_all):
...
+                num_modes_sorted = filter_metric_sorted.sizes["mode_index"]
+                if keep_mask.sum() < num_modes_sorted:
+                    keep_inds = np.where(keep_mask)[0]
+            elif isinstance(keep_modes, int):
+                if keep_mask.sum() < keep_modes:
+                    log.warning(
+                        f"Filtering with 'keep_modes={keep_modes}' keeps "
+                        f"more than the '{keep_mask.sum()}' modes which pass the filter. Consider "
+                        "relaxing the filter threshold or changing 'keep_modes'."
+                    )
```
[the remark this answers is not in the document]
Comment (caseyflex) [reply tidy3d/components/data/monitor_data.py]: the message was correct but the new one is clearer

**Comment by caseyflex** [comment]
@greptile

**Comment by caseyflex** [comment]
@greptile

File: tidy3d/components/data/monitor_data.py [hunk 6]
```diff
@@ -2404,31 +2578,94 @@ def _order_indices(indices, vals_all):
...
+                if keep_mask.sum() < num_modes_sorted:
+                    keep_inds = np.where(keep_mask)[0]
+            elif isinstance(keep_modes, int):
+                if keep_mask is not None and keep_mask.sum() < keep_modes:
+                    log.warning(
+                        f"'keep_modes={keep_modes}' requests {keep_modes} modes, but only "
+                        f"{keep_mask.sum()} modes pass the filter. All {keep_modes} modes will be kept. "
+                        "Consider relaxing the filter threshold or lowering 'keep_modes'."
+                    )
+                if keep_modes > num_modes:
```
[the remark this answers is not in the document]
Comment (caseyflex) [reply tidy3d/components/data/monitor_data.py:2650]: yes, when sort_spec is passed directly to this function

**Comment by caseyflex** [comment]
@greptile
